service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,337 @@ package sysworkflow
 	"go.uber.org/cadence/workflow"
-	"go.uber.org/zap"
 	"time"
 )
 
+var (
+	errArchivalUploadActivityGetDomain           = cadence.NewCustomError(errArchivalUploadActivityGetDomainStr)
+	errArchivalUploadActivityNextBlob            = cadence.NewCustomError(errArchivalUploadActivityNextBlobStr)
+	errArchivalUploadActivityConstructKey        = cadence.NewCustomError(errArchivalUploadActivityConstructKeyStr)
+	errArchivalUploadActivityBlobExists          = cadence.NewCustomError(errArchivalUploadActivityBlobExistsStr)
+	errArchivalUploadActivityMarshalBlob         = cadence.NewCustomError(errArchivalUploadActivityMarshalBlobStr)
+	errArchivalUploadActivityConvertHeaderToTags = cadence.NewCustomError(errArchivalUploadActivityConvertHeaderToTagsStr)
+	errArchivalUploadActivityWrapBlob            = cadence.NewCustomError(errArchivalUploadActivityWrapBlobStr)
+	errArchivalUploadActivityUploadBlob          = cadence.NewCustomError(errArchivalUploadActivityUploadBlobStr)
+	errDeleteHistoryActivityDeleteFromV2         = cadence.NewCustomError(errDeleteHistoryActivityDeleteFromV2Str)
+	errDeleteHistoryActivityDeleteFromV1         = cadence.NewCustomError(errDeleteHistoryActivityDeleteFromV1Str)
+)
+
 // SystemWorkflow is the system workflow code
 func SystemWorkflow(ctx workflow.Context) error {
-	id := workflow.GetInfo(ctx).WorkflowExecution.ID
-	logger := workflow.GetLogger(ctx)
-	scope := workflow.GetMetricsScope(ctx).Tagged(map[string]string{SystemWorkflowIDTag: id})
-	ch := workflow.GetSignalChannel(ctx, SignalName)
-
-	logger.Info("started new system workflow")
+	sysWorkflowInfo := workflow.GetInfo(ctx)
+	logger := globalLogger.WithFields(bark.Fields{
+		logging.TagWorkflowExecutionID: sysWorkflowInfo.WorkflowExecution.ID,
+		logging.TagWorkflowRunID:       sysWorkflowInfo.WorkflowExecution.RunID,
+	})
+	logger.Info("started system workflow")
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerWorkflowStarted)
+	sw := globalMetricsClient.StartTimer(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNewLatency)
+	ch := workflow.GetSignalChannel(ctx, signalName)
 	signalsHandled := 0
-	for ; signalsHandled < SignalsUntilContinueAsNew; signalsHandled++ {
+	for ; signalsHandled < signalsUntilContinueAsNew; signalsHandled++ {
 		var signal signal
 		if more := ch.Receive(ctx, &signal); !more {
-			scope.Counter(ChannelClosedUnexpectedlyError).Inc(1)
-			logger.Error("cadence channel was unexpectedly closed")
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
 	}
 
 	for {
 		var signal signal
 		if ok := ch.ReceiveAsync(&signal); !ok {
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
```

Comment:
This could potentially cause you to miss signals.  It is unsafe to do any processing of signals after draining the signal pump.  I think the safest thing to do is pass all those signals to new run for processing.

Reply:
Let's also chat about this one over zoom.